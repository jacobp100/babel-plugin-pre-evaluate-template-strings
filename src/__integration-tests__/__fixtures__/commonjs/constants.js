// eslint-disable-next-line
module.exports = {
  fontSize: '14px',
  unitless: {
    fontSize: 14,
  },
};
