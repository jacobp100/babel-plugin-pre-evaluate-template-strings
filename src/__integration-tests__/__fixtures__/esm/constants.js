export default {
  fontSize: '14px',
};
