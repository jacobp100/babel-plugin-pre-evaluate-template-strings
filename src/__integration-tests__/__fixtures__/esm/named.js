export const base = {
  fontSize: '24px',
};

export const primary = {
  fontSize: '36px',
};
