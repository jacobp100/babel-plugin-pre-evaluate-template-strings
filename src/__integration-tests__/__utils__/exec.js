/* @flow */

import childProcess from 'child_process';
import path from 'path';

export function transpile(
  source: string,
  pluginOptions: any = { cache: false, extract: false },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_TRANSPILE_BIN = path.join(__dirname, './transpile.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_TRANSPILE_BIN} '${JSON.stringify(
      source
    )}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [transpiled, rawStyles] = stdout.toString().split('\n');
  const styles = JSON.parse(rawStyles || '{}');
  return {
    code: JSON.parse(transpiled || '""'),
    styles,
    getCSSForClassName(className: string) {
      return styles[`.${className}`];
    },
  };
}

export function extract(
  source: string,
  pluginOptions: any = { cache: false, extract: true },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_EXTRACT_BIN = path.join(__dirname, './extract.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_EXTRACT_BIN} '${JSON.stringify(source)}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [data, filenames, transpiled] = stdout.toString().split('\n');
  return {
    data: JSON.parse(data || '""'),
    filenames: JSON.parse(filenames || '[]'),
    transpiled: JSON.parse(transpiled || '""'),
  };
}
