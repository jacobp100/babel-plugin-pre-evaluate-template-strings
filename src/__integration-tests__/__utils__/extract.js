const fs = require('fs');

const writeFileSync = fs.writeFileSync;

let buffer = '';
const filenames = [];

fs.writeFileSync = (filename, data, opts) => {
  if (/\.css$/.test(filename)) {
    filenames.push(filename);
    buffer += data;
    return;
  }
  writeFileSync(filename, data, opts);
};

const transpile = require('./transpile');

const code = transpile(
  JSON.parse(process.argv[2]),
  JSON.parse(process.argv[3])
);

console.log(JSON.stringify(buffer));
console.log(JSON.stringify(filenames));
console.log(JSON.stringify(code || ''));
