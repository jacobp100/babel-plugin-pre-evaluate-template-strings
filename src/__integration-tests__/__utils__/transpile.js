const babel = require('babel-core');
const path = require('path');
const dedent = require('dedent');

const CSS_PATH = require.resolve(path.join(process.cwd(), 'build/css'));
const SHEET_PATH = require.resolve(path.join(process.cwd(), 'build/sheet'));
const PREVAL_PLUGIN_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/preval-extract')
);
const MODULE_SYSTEM_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/lib/moduleSystem')
);

function transpile(source, options = { pluginOptions: {}, babelOptions: {} }) {
  return babel.transform(
    dedent`
      import css from '${CSS_PATH}';

      ${source}
    `,
    Object.assign(
      {
        presets: ['env', 'stage-2', 'react'],
        plugins: [[PREVAL_PLUGIN_PATH, options.pluginOptions]],
        sourceMaps: false,
      },
      options.babelOptions
    )
  ).code;
}

module.exports = transpile;

if (require.main && module.filename === require.main.filename) {
  const code = transpile(
    JSON.parse(process.argv[2]),
    JSON.parse(process.argv[3])
  );

  if (code) {
    console.log(JSON.stringify(code));

    // eslint-disable-next-line global-require
    const getCachedModule = require(MODULE_SYSTEM_PATH).getCachedModule;
    const sheet = getCachedModule(SHEET_PATH);
    const styles = sheet
      ? getCachedModule(SHEET_PATH).exports.default.styles()
      : {};
    console.log(JSON.stringify(styles || ''));
  }
}
