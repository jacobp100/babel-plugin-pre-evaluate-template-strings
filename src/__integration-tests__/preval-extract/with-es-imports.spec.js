/* eslint-disable no-template-curly-in-string */
/* @flow */

import dedent from 'dedent';
import { transpile } from '../__utils__/exec';

describe('preval-extract babel plugin with ES imports', () => {
  it('should preval default export', () => {
    const { code, getCSSForClassName } = transpile(dedent`
      import constants from "./src/babel/__integration-tests__/__fixtures__/esm/constants.js";

      const header = css\`
        font-size: ${'${constants.fontSize}'};
      \`;
      `);

    const match = /header = \/\*.+\*\/"(_header__[a-z0-9]+)"/g.exec(code);
    expect(match).not.toBeNull();
    const css = getCSSForClassName(match[1]);
    expect(css).toMatch('font-size: 14px');
    expect(css).toMatchSnapshot();
  });

  it('should preval named imports', () => {
    const { code, getCSSForClassName } = transpile(dedent`
      import { base, primary } from "./src/babel/__integration-tests__/__fixtures__/esm/named.js";

      const header = css\`
        font-size: ${'${primary.fontSize}'};
      \`;

      const body = css\`
        font-size: ${'${base.fontSize}'};
      \`;
      `);

    const headerMatch = /header = \/\*.+\*\/"(_header__[a-z0-9]+)"/g.exec(code);
    const bodyMatch = /body = \/\*.+\*\/"(_body__[a-z0-9]+)"/g.exec(code);
    expect(headerMatch).not.toBeNull();
    expect(bodyMatch).not.toBeNull();
    const headerStyles = getCSSForClassName(headerMatch[1]);
    const bodyStyles = getCSSForClassName(bodyMatch[1]);
    expect(headerStyles).toMatch('font-size: 36px');
    expect(headerStyles).toMatchSnapshot();
    expect(bodyStyles).toMatch('font-size: 24px');
    expect(bodyStyles).toMatchSnapshot();
  });

  it('should preval imported module tree with constants', () => {
    const { code, getCSSForClassName } = transpile(dedent`
      import constants from "./src/babel/__integration-tests__/__fixtures__/esm/deep.js";

      const header = css\`
        font-size: ${'${constants.fontSize}'};
      \`;
      `);

    const match = /header = \/\*.+\*\/"(_header__[a-z0-9]+)"/g.exec(code);
    expect(match).not.toBeNull();
    const css = getCSSForClassName(match[1]);
    expect(css).toMatch('font-size: 28px');
    expect(css).toMatchSnapshot();
  });
});
