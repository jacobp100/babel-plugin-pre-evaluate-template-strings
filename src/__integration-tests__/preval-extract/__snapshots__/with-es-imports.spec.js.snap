// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval-extract babel plugin with ES imports should preval default export 1`] = `"font-size: 14px;"`;

exports[`preval-extract babel plugin with ES imports should preval imported module tree with constants 1`] = `"font-size: 28px;"`;

exports[`preval-extract babel plugin with ES imports should preval named imports 1`] = `"font-size: 36px;"`;

exports[`preval-extract babel plugin with ES imports should preval named imports 2`] = `"font-size: 24px;"`;
